### ISSUE TITLE
Mempool size protection improvements for current flooding style

### ISSUE THREAD

**Issue by Beanow** [bitcoinxt/bitcoinxt#61]
@gavinandresen wrote in #56:

```
Implement a simple strategy to keep the memory pool from consuming
all of memory, by limiting the number of tranasctions stored in
the memory pool.
```

Today there was a flooding of the mempool that didn't rely on the amount of transactions but instead [used 500 inputs](https://tradeblock.com/bitcoin/address/1aaf64VJGMuYuRBMmM9y2s2pPADFpuqDE) to bloat each transaction to ~73kb and using a 1 satoshi/byte fee to have it linger around long. Resulting in peaks of ~48mb of mempool even though the backlog didn't go far above 3500 TX (even in between 50min blocks). For reference, plenty of the 1MB blocks saw >2200 TX ([example](https://tradeblock.com/bitcoin/block/0000000000000000088cd533b4f34434d60f3bbb96910dfec6880d6ed304ed76)).

Had this "tester" decided to generate these at a faster rate this could have incurred a lot more RAM usage while keeping TX count small. And a smaller bad TX : good TX ratio means higher odds the random removal will find a legitimate TX.

Perhaps an additional setting that limits the maximum mempool total size should be added to trigger a random remove besides just `-maxmempooltx`. For example `-maxmempoolsize=4096` to set a 4096KB limit.

Note that blockchain.info instead decided to reject the TX completely:

```
Transaction rejected by our node. Reason:
The Maximum number of inputs in a single transaction is 200
```

But that is an option I believe would need to be part of the protocol to sanely enforce.

Instead, perhaps large input TX like these should be written to disk after we're done propagating it to not rely on RAM as it could be a while before these are taken up into a block. Recent fee estimation code could help decide if it's worth the I/O too, writing out large TX with low odds of making it into the next block.

**Comment by dgenr8** [comment]
Perhaps ejection should pick a random *byte* out of the mempool, and eject the tx it belongs to.  Fun implementation challenge.

**Comment by mikehearn** [comment]
You're right, but bear in mind such transactions consume UTXOs very fast. You can't just chain them forever.

Better algorithms to pick a random tx are certainly desirable. As the actual constraint is RAM, configuring by byte-size rather than num-txns is clearly superior. Picking a random byte and mapping it to the transaction is a good suggestion, but, would take more code and probably have worse performance. The additional CPU cost may not matter though.

The code in 0.11B represents what Gavin could implement quickly enough rather than a theoretical ideal. We can and should upgrade it in future releases.

**Comment by gavinandresen** [comment]
On Wed, Sep 9, 2015 at 9:16 AM, Mike Hearn <notifications@github.com> wrote:

> The code in 0.11B represents what Gavin could implement quickly enough
> rather than a theoretical ideal. We can and should upgrade it in future
> releases.


Yes; see https://github.com/bitcoin/bitcoin/pull/6654/files for the
direction Core is going, and where you end up once you start down the path
of "lets be smarter about what we eject".

-- 
--
Gavin Andresen

**Comment by Beanow** [comment]
@gavinandresen their code looks like a @#!% to implement across all wallets to avoid it.

@mikehearn agreed this issue shouldn't hold up the current release.
I figured I would start the improvements discussion here since we have a real-life sample.
I'll rename to clarify.

**Comment by gavinandresen** [comment]
On the original topic of large, very-low-fee transactions:

They'll be rate-limited by this code:
```
        // Continuously rate-limit free (really, very-low-fee) transactions
        // This mitigates 'penny-flooding' -- sending thousands of free transactions just to
        // be annoying or make others' transactions take longer to confirm.
        if (fLimitFree && nFees < ::minRelayTxFee.GetFee(nSize))
        {
            static double dFreeCount;
            static int64_t nLastFreeTime;
            static int64_t nFreeLimit = GetArg("-limitfreerelay", 15);

            // At default rate it would take over a month to fill 1GB
            if (RateLimitExceeded(dFreeCount, nLastFreeTime, nFreeLimit, nSize))
                return state.DoS(0, error("AcceptToMemoryPool : free transaction rejected by rate limiter"),
                                 REJECT_INSUFFICIENTFEE, "rate limited free transaction");

            LogPrint("mempool", "Rate limit dFreeCount: %g => %g\n", dFreeCount, dFreeCount+nSize);
        }
```

... so I'm not sure there is anything to fix; a lower default for -limitfreerelay on lower-memory systems would make sense.

**Comment by Beanow** [comment]
@gavinandresen interesting, I hadn't thought of this code since it didn't kick in. It seems in this attack they precisely avoided the [default value](https://github.com/bitcoinxt/bitcoinxt/blob/master/src/main.cpp#L71) with 1000 satoshis per 1000 bytes.
Making `nFees == ::minRelayTxFee.GetFee(nSize)` and not triggering this if statement.
They also avoided the code for detecting dust by having this high amount of inputs.

```
Target: 1aaf64VJGMuYuRBMmM9y2s2pPADFpuqDE 0.00426459 BTC
Fee:                                      0.00073541 BTC
Size:                                          73541 B
```

Would have been interesting to see this in action by increasing for example to `-minrelaytxfee=0.000011` rather than the default `0.00001`.

**Comment by mikehearn** [comment]
Yes, the issue with these sorts of hard-coded defaults is DoS attackers can just go around them. I'm not a big fan of the rate limiter anyway. The rate limiter is why no wallets send free transactions in practice even if theoretically they could: it's just too unreliable (wallets are incentivised to minimise their fees).

I still like Tom's idea of random-byte. The resource that's limited is measured in megabytes rather than transactions, so it makes sense to set the limit in megabytes and not transactions. I wonder if it's even that difficult. We already track mempool total size. Pick a random value within that range, then iterate over the map accumulating a counter until it goes over the random value. That's essentially the same as picking a random byte. It's O(N/2) instead of O(log N) but as the values we're talking about are only small that probably doesn't matter.

**Comment by Beanow** [comment]
Hmm, I'm having some doubts about the byte based searching. It would make it more likely that large TX are picked out. Legitimate use-cases (such as paying someone 5 BTC with a large amount of mining pool payouts as input) would be at risk of being discriminated against and experiencing bigger propagation issues than a small (2 in, 2 out, 400 bytes) TX would.

So in other words, I think in theory being blind to the TX size would make it more likely a TX is propagated and preserved across the network as a whole.

Wallets might implement dodgy workarounds to protect their users from being discriminated against when sending large TX. Like to split them up and wait for the mempool to shrink before sending the next ones. Or to keep re-sending. Which could make matters complicated, more expensive and cost more bandwidth.

**Comment by dgenr8** [comment]
@gavinandresen is right, even if preferring bigger txes is better, it's just one of quite a few interesting metrics, the balancing of which tends to bog down this discussion.

The random solution is very neutral.  I really like the fact that it prioritizes using the fee-based confirmation estimate, and NOT the fee itself.  That lines up with the node's interest (clearing out mempool quickly) without completely siding with miners (highest possible fee, period).

**Comment by Beanow** [comment]
@gavinandresen attacks have been ongoing and I found `-minrelaytxfee=0.000012` to be very effective. Though the majority is not rejected through the rate limiting, but priority requirement.

```
2015-09-10 00:21:00 5a1ac0d4513e3af8894cea53a1313cb04b587a0dc3697c6f9bcf7a4055cb5a7e from peer=42 /Satoshi:0.10.2/ was not accepted into the memory pool: insufficient priority
```
[^above TX](https://tradeblock.com/bitcoin/tx/5a1ac0d4513e3af8894cea53a1313cb04b587a0dc3697c6f9bcf7a4055cb5a7e)

It seems as though they also attempted to game the "dust" code because it's based on the same default value. Spamming many of [these TX](https://tradeblock.com/bitcoin/tx/9c2eb0ef3e3140f477a49aff74918a835edd3e932b147d59c889476b7702df2d) that are slightly above the 1/3 ratio. Raising `minrelaytxfee` catches all of those too.

**Comment by Beanow** [comment]
![TX backlog count](http://ss.oscp.info/2015-09-10_15-46-12.png)
![TX backlog size](http://ss.oscp.info/2015-09-10_15-46-34.png)

Given the above graphs,

I do think that supporting `-maxmempoolsize` would provide a valuable and (more?) reliable metric when it comes to keeping your node from running out of RAM. While keeping the same method of removing the TX (randomly, by TX index).

Like @mikehearn said. "The resource that's limited is measured in megabytes rather than transactions, so it makes sense to set the limit in megabytes and not transactions."

**Comment by gavinandresen** [comment]
I agree, modifying the code to:

a) Limit by #bytes in mempool instead of #transactions (mempool.GetTotalTxSize already keeps track, can evict transactions until size is below limit)
b) Idea from mailing list: pick n random transactions (n == 2 or 3 or 4), evict one with lowest fee/kb
... would make it better.

I'll be traveling for the next ten days or so, if anybody else wants to tackle those changes they should be easy.

**Comment by Beanow** [comment]
To update on the chart:
![TX size](http://ss.oscp.info/2015-09-11_01-50-33.png)
![TX count](http://ss.oscp.info/2015-09-11_01-51-02.png)

In the end the 50k default TX limit triggered and both limit types seem to be viable at reducing RAM use. The "test" employed different ways to bloat the mempool and TX count and depending on the type of scheme used one or the other will detect a DoS first.

My question being, @gavinandresen do you think #transactions limit should be **replaced** by #bytes? Or provide them both?

+1 for the pick n random to evict lowest fee/kb suggestion [link to post on mailing list](https://groups.google.com/d/msg/bitcoin-xt/3YwHm56RuLw/ItCJEm16AwAJ).

Should the N value be configurable and with which constraints? 1-4? 1-mempoolsize?

**Comment by dgenr8** [comment]
Outdated.  XT implemented child-pays-for-ancestors in #90.
